tmol/pose/constraint_set.py

```diff
@@ -0,0 +1,115 @@
+import attr
+import torch
+
+from tmol.types.torch import Tensor
+from tmol.chemical.restypes import RefinedResidueType
+from tmol.pose.packed_block_types import PackedBlockTypes
+
+
+class ConstraintSet:
+    """ """
+
+    MAX_N_ATOMS = 4
+
+    device: torch.device
+    # constraint_index x (type, pose_index)
+    constraint_function_inds: Tensor[torch.int][:]
+    constraint_atom: Tensor[torch.int][:, 4, 3]
+    constraint_params: Tensor[torch.float32][:, :]
+    constraint_functions = []
+
+    def __init__(self, device):
+        self.constraint_function_inds = torch.full(
+            (0,), 0, dtype=torch.int32, device=device
+        )
+        self.constraint_atoms = torch.full(
+            (0, self.MAX_N_ATOMS, 3), 0, dtype=torch.int32, device=device
+        )
+        self.constraint_params = torch.full(
+            (0, 1), 0, dtype=torch.float32, device=device
+        )
+        self.device = device
+
+    #################### PROPERTIES #####################
+
+    def add_constraints_to_all_poses(self, fn, nposes, atom_indices, params=None):
+        self.add_constraints(fn, atom_indices, params, nposes=nposes)
+
+    def add_constraints(self, fn, atom_indices, params=None, nposes=0):
```

Comment:
We have to pass in the number of poses in order to replicate a copy of each constraint to all poses, since this constraint_set doesn't have a reference to the pose. 

I'm thinking more and more that these functions should be going through pose stack (and then maybe managed internally in the pose_stack by a constraint_set class?). The constraints are fundamentally attached to the pose_stack anyway, especially if you get the atom names from the pose stack's block types like in the test cases.

Reply:
See above